### ISSUE TITLE
Allow users to encrypt wallet / wallet encryption/decryption user flow

### ISSUE THREAD

**Issue by tzarebczan** [lbryio/lbry-desktop#762]
## The Issue
Once @akinwale is done with the protocol side of this which allows wallet encryption, we need to have a discussion about what this should look/work like from a user experience standpoint. 

Some quick points:
1) This needs to be as simple as possible - we'll have many non-crypto  / non-tech users
2) We need to be explicit about not being able to recover the wallet if the encryption password is lost
3) Do we allow some type of pin which allows wallet to be unlocked via the auth token? 
4) When is the wallet unlocked? 
5) Can it be re-locked? 
6) Possible to integrate some type of 2FA? 

I'd like to hear @kauffj 's thoughts on what this user flow should look like too :)

**Comment by kauffj** [comment]
IMO this is something we ought to avoid introducing on first-run unless we have to. If it must be introduced on first-run, do it in a way that a user can put it off or do it later.

When the user is first running the software, they don't have any credits, or have very few. Being told you must write something down and if you forget it you lose your credits forever is a very intimidating introduction to something new.

Instead, I'd propose we introduce this either once the wallet crosses a certain threshold, or only if the user specifically indicates a desire for this type of security.

**Comment by tzarebczan** [comment]
@akinwale feel free to use this issue or close/re-open a new one.

**Comment by akinwale** [comment]
## Wallet Encryption Commands
There are three new daemon CLI and JSON RPC commands related to wallet encryption.
- wallet_encrypt
- wallet_unlock
- wallet_decrypt

### wallet_encrypt
Encrypts the wallet private key, thereby requiring having to unlock the wallet before transactions can be sent from the wallet. The `master_private_keys` field in the default_wallet file will be updated with the encrypted private key string. After a wallet has been encrypted, you will need to use the wallet_unlock command to allow the wallet to be used. An encrypted wallet also needs to be unlocked upon daemon start up. There is a log entry during the daemon start up process which reads "Waiting for wallet password". Using `wallet_unlock` with the correct password at this point will allow the daemon to launch successfully. 

#### CLI
`lbrynet-cli wallet_encrypt <new_password>`

#### JSON RPC
`curl -X POST -d '{"method":"wallet_encrypt", "params":{"new_password":"<new_password>"}}' http://localhost:5279`


### wallet_unlock
Unlocks the wallet if the specified password matches the password that was used to encrypt the wallet. As stated above, the wallet needs to be unlocked to send transactions, or in order to allow the to daemon start up completely.

#### CLI
`lbrynet-cli wallet_unlock <password>`

#### JSON RPC
`curl -X POST -d '{"method":"wallet_unlock", "params":{"password":"<password>"}}' http://localhost:5279`


### wallet_decrypt
Removes encryption from the wallet private key. The `master_private_keys` field in the default_wallet file will be updated with plain private key string. An encrypted wallet can only be decrypted after it has been unlocked.

#### CLI
`lbrynet-cli wallet_decrypt`

#### JSON RPC
`curl -X POST -d '{"method":"wallet_decrypt", "params":{}}' http://localhost:5279`


## How to determine if a wallet is encrypted?
Running the `status` command will return a number of fields, including `wallet_is_encrypted` which is boolean, with a value of `true` if the wallet is encrypted, or `false` otherwise.

#### CLI
`lbrynet-cli status`

#### JSON RPC
`curl -X POST -d '{"method":"status", "params":{}}' http://localhost:5279`

**Comment by seanyesmunt** [comment]
Looks good to me. I'm guessing we will just have a check on app startup and, if its encrypted, show a password field to unlock before loading anything else?

**Comment by akinwale** [comment]
Yes, looks like it. You can run the status command on app startup (if the daemon is being started by the app) to check if the wallet is encrypted. If the wallet is waiting for a password, you also get
```
"startup_status": {
    "code": "loading_wallet",
    "message": "Catching up with the blockchain"
  }
```
in the response.

**Comment by LavRadis** [comment]
I've been testing this and at one point the console returned "true" when I did the unlock command with a false password. Not sure if I followed the correct protocol. I've attached a .txt of that session. The return is on line 88. 
[lbry encrypt test 2.txt](https://github.com/lbryio/lbry-app/files/1620110/lbry.encrypt.test.2.txt)

**Comment by tzarebczan** [comment]
@akinwale per our discussion on Slack, we should have a wallet_lock or similar which will prevent transactions from being signed. The use case would be users running the daemon for hosting purposes, but they aren't necessarily transacting. 

This would require some overhead on the app side - would need a way to allow lock and then unlock again - prompt user to do so when a transaction needs to occur.

**Comment by tzarebczan** [comment]
@akinwale / @jackrobison 

We should also be encrypting channel certificate private keys during this process as well, right? I ran through a quick test and it doesn't look like we are.

**Comment by tzarebczan** [comment]
@akinwale did you have a chance to review the comments from @LavRadis per our discussion?

Also, can you comment on using the keyring to store the encryption keys?

**Comment by tzarebczan** [comment]
@hackrush01 think you connected this to the wrong PR

**Comment by skhameneh** [comment]
Working on this

**Comment by tzarebczan** [comment]
use keyring setting: https://github.com/lbryio/lbry/blob/6952c2c07e099a13afa6458baa47fa2ea202340a/lbrynet/conf.py#L297

**Comment by alyssaoc** [comment]
a duplicate of #1097, this issue closed in of another issue
